### ISSUE TITLE
CIP-0030 | Add JSON Web Token generation with domain name validation

### ISSUE THREAD

**Issue by makerare** [cardano-foundation/CIPs#448]
Currently, any hybrid dApp can authenticate its users using CIP-0030 `signData` function. An implementation of an equivalent of JSON Web Token (JWT) is [available here](https://github.com/pyropy/web3-cardano-token). 

The problem relying in this method is its vulnerability to a phishing attack: a malicious website could forge a JWT made for authentication to another website, make the user sign it, and act as a man in the middle.
Example: User fetches malicious jpgs.store, signs a message containing "Connect to jpg.store". Signed message can the be used to access user's account on jpg.store Hybrid dApp (which is critical in this case as you can manage NFT collections and their royalty address from their interface).

A solution could be to implement JW3T signature directly in the wallet API.
```
Currently, any hybrid dApp can authenticate its users using CIP-0030 `signData` function. An implementation of an equivalent of JSON Web Token (JWT) is [available here](https://github.com/pyropy/web3-cardano-token). 

The problem relying in this method is its vulnerability to a phishing attack: a malicious website could forge a JWT made for authentication to another website, make the user sign it, and act as a man in the middle.
Example: User fetches malicious jpgs.store, signs a message containing "Connect to jpg.store". Signed message can the be used to access user's account on jpg.store Hybrid dApp (which is critical in this case as you can manage NFT collections and their royalty address from their interface).

A solution could be to implement JW3T signature directly in the wallet API.
```
import Web3Token from 'web3-cardano-token/dist/browser';

walletApi.signJW3T = async function (
  address: string,
  body: object, 
  expires_in: string
): string {
  /*
    Forges a Web3Token using the 'web3-cardano-token' standard but with addition of domain of the current website:
  */
    return await Web3Token.sign(
      {
         signer: msg => walletApi.signData(address, toHex(msg)),
         expire_in: expires_in,
         body: body,
         domain: 'jpg.store' // Wallet API inserts here the current domain
      }
    );
}
```
`walletApi.signData` also need to be display a warning sign when it is called to sign a message with JW3T format but not the good domain.

To verify the token from the backend:
```
const Web3Token = require('web3-cardano-token/dist/node');

const token = req.headers['Authorization']
const { address, body, domain } = await Web3Token.verify(token);

// Check if domain matches
```

This standard is very straight forward to implement, as `web3-cardano-token` package provides it with just the addition of the domain key (which could also simply go inside the body object).

**Comment by rphair** [comment]
@makerare if you're thinking of writing a CIP for this, it sounds like it might be a good idea to propose.  We have a new convention under review for extensions to CIP-0030: https://github.com/cardano-foundation/CIPs/pull/446

If you're suggesting this be added to CIP-0030 under all circumstances then say so here & I'll tag some people involved in the history of CIP-0030 to help discuss that.  (You are welcome to do this as well.)

**Comment by makerare** [comment]
Yes I would appreciate this as I don't know who to reach, so they can arbitrate as they would know what makes the most sense, thank you. I'd be glad to write it !

**Comment by rphair** [comment]
cc (may be interested in issue discussion of CIP update or new CIP for this): @KtorZ @SebastienGllmt @alessandrokonrad @rooooooooob @ehanoc @refi93 @MarcelKlammer @Scitz0

**Comment by makerare** [comment]
After some thinking, i think what makes the most sense is to use [JWT standard](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.3), and write a library such as `web3-cardano-token` for generating data to sign but implementing the exact standard.

Then there would be no need to write a new function for the API. It should just be specified in the CIP 0030 signData description that wallet should check if signed data corresponds to a JW3T, if that's the case, wallet could have a special interface for displaying this to the user.

As the standard contains an "aud" (audience) field. If JW3T signature attempte contains "uri:domain.com" different from the current domain wallet should warn the user the possible phishing attempt.

**Comment by rphair** [comment]
@pa-long I'm closing this since 3.5 years have gone by (though honestly I'm only just checking today after a long lapse) since your offer to submit a PR on CIP-0030.  Please note in the meantime we have adopted (and mandated in CIP-0030 itself, as I recall) that further updates to the interface should be in the form of new CIPs which define extensions.

It also appears that your particular implementation idea has been absorbed into this linked issue & should be continued at the library level:
* https://github.com/pyropy/web3-cardano-token/issues/7

So it does feel safe to close this but please if this is something you still think needs to be done the submit it next time _as a PR_ related to CIP-0030; then editors would be able to circulate it promptly for review in the wallet dev community.
